systems/controllers/QPCommon.cpp

```diff
@@ -564,11 +689,24 @@ int setupAndSolveQP(NewQPControllerData *pdata, std::shared_ptr<drake::lcmt_qp_c
   Vector6d Jbdotv;	
   for (int i=0; i<desired_body_accelerations.size(); i++) {
     if (desired_body_accelerations[i].weight < 0) { // negative implies constraint
-      if (!inSupport(active_supports,desired_body_accelerations[i].body_id0)) {
-        pdata->r->forwardJac(desired_body_accelerations[i].body_id0,orig1,1,Jb);
-        auto Jdot_times_v = pdata->r->forwardJacDotTimesV(orig,desired_body_accelerations[i].body_id0,0,1,0);
-        Jbdotv.noalias() = Jdot_times_v.value();
-
+      int body_id0;
+      if (desired_body_accelerations[i].body_or_frame_id0 >= 0) {
+        body_id0 = desired_body_accelerations[i].body_or_frame_id0;
+      } else {
+        int frame_ind = -(desired_body_accelerations[i].body_or_frame_id0 + 2);
+        body_id0 = pdata->r->frames[frame_ind].body_ind;
+      }
+      if (desired_body_accelerations[i].control_pose_when_in_contact || !inSupport(active_supports,body_id0)) {
+        auto J_geometric = pdata->r->geometricJacobian<double>(0,desired_body_accelerations[i].body_or_frame_id0,desired_body_accelerations[i].body_or_frame_id0,0,true,(std::vector<int>*)nullptr);
+        auto J_geometric_dot_times_v = pdata->r->geometricJacobianDotTimesV<double>(0,desired_body_accelerations[i].body_or_frame_id0,desired_body_accelerations[i].body_or_frame_id0,0); 
+        Matrix<double,6,-1> Jb_compact = J_geometric.value();
+        Jb = pdata->r->compactToFull<Matrix<double,6,-1>>(Jb_compact,desired_body_accelerations[i].body_path.joint_path,true);
```

Comment:
Let's keep using Dynamic instead of -1. Who knows, they could change what Eigen::Dynamic is in a subsequent release. Also clearer.